client/notices.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+type NotifyOptions struct {
+	// Key is the client notice's key. Must be in "domain.com/key" format.
+	Key string
+
+	// RepeatAfter, if provided, allows the notice to repeat after this duration.
+	RepeatAfter time.Duration
+
+	// Data are optional key=value pairs for this occurrence of the notice.
+	Data map[string]string
+}
+
+// Notify records an occurrence of a "client" notice with the specified options,
+// returning the notice ID.
+func (client *Client) Notify(opts *NotifyOptions) (string, error) {
+	var payload = struct {
+		Action      string            `json:"action"`
+		Type        string            `json:"type"`
+		Key         string            `json:"key"`
+		RepeatAfter string            `json:"repeat-after,omitempty"`
+		Data        map[string]string `json:"data,omitempty"`
+	}{
+		Action: "add",
+		Type:   "client",
+		Key:    opts.Key,
+		Data:   opts.Data,
+	}
+	if opts.RepeatAfter != 0 {
+		payload.RepeatAfter = opts.RepeatAfter.String()
+	}
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(&payload); err != nil {
+		return "", err
+	}
+
+	result := struct {
+		ID string `json:"id"`
+	}{}
+	_, err := client.doSync("POST", "/v1/notices", nil, nil, &body, &result)
+	if err != nil {
+		return "", err
+	}
+	return result.ID, err
+}
+
+type NoticesOptions struct {
+	// Type, if set, includes only notices of this type.
+	Type NoticeType
+	// Key, if set, includes only notices with this key.
+	Key string
+	// After, if set, includes only notices that were last repeated after this time.
+	After time.Time
+}
+
+// Notice is a notification whose identity is the combination of Type and Key
+// that has occurred Occurrences number of times.
+type Notice struct {
```

Comment:
is this not exactly the same struct as in the state code? If so, why duplicate it here? Can't we just reuse the same struct? (Maybe put it in a separate package if we don't want `client` to depend on `state`).

Ditto for `NoticeType` and `jsonNotice`.

Reply:
We definitely don't want `client` to depend on `state`. We could put it in a common package as you say, but I think it's cleaner to just duplicate it -- it keeps the separation crystal clear, and it allows the state marshalling and API to evolve at a different rate than the client. For example, we might add internal notice types that we don't want to expose to the client, or new `Notice` fields that are only internally-relevant.

In this specific case, they're not quite the same ... if we did have a common package, it would have to expose the `JsonNotice` type for unmarshalling, but the real `Notice` type is actually different between the two (the `state` version intentionally doesn't export any fields, the `client` version intentionally exports all fields). We could probably still share some stuff, but it quickly gets messy and you have to find common APIs that work.

And once again, the precedent here is to not share code between server and client. (If the client were in a separate repo, we almost certainly wouldn't share it.)

I've often found the Go proverb to be true: "A little copying is better than a little dependency".